group/assignment5/DirectedGraphNode.java

```diff
@@ -0,0 +1,138 @@
+    private ArrayList<DirectedGraphNode<T>> parents;
+    protected ArrayList<DirectedGraphNode<T>> children;
+
+    public final static int BRANCHING_FACTOR = Integer.MAX_VALUE; // makes no assumption on the size of the alphabet
+
+    public DirectedGraphNode(T key) {
+        this.key = key;
+        children = new ArrayList<DirectedGraphNode<T>>();
+        parents = new ArrayList<DirectedGraphNode<T>>();
+    }
+
+
+    /**
+     * Checks whether the subtree starting from this DirectedGraphNode contains a node
+     * with specific key
+     *
+     * @param key, a T object, node to search in the tree induced by this node
+     * @return the node corresponding to the key if it exists in the subtree, null
+     *         otherwise
+     */
+    public DirectedGraphNode<T> findNode(T key) {
+        if(key==null)
+            throw new IllegalArgumentException("The argument is null!");
+        return findNodeRecursive(this, key);
+    }
+
+    /**
+     * Looks for a child node with a specific key in the direct children of the current node.
+     * @param childKey a Character, the key of the child to search
+     * @return the DirectedGraphNode corresponding to the child with key childKey if it exists, null otherwise
+     */
+    public TrieNode findChild(T childKey) {
+        if(childKey==null)
+            throw new IllegalArgumentException("The argument is null!");
+        for(DirectedGraphNode<T> child : this.children) {
+            if (child.getKey() == childKey)
+                return (TrieNode) child;
+        }
+        return null;
+    }
+
+    private DirectedGraphNode<T> findNodeRecursive(DirectedGraphNode<T> current, T key) {
+        if (current.key.equals(key))
+            return current;
+
+        DirectedGraphNode<T> found = null;
+        for (DirectedGraphNode<T> child : current.getChildren()) {
+            found = findNodeRecursive(child, key);
+            if (found != null) {
+                return found;
+            }
+        }
+        return found;
+    }
+
+    /**
+     * Checks if this node is equals to the DirectedGraphNode node.
+     *
+     * @param node
+     *            a DirectedGraphNode to be compared against this node
+     * @return true if the this node and the TreeNode node have the same key
+     *         and the same parent.
+     */
+    public boolean equals(DirectedGraphNode<T> node) {
+        if(node==null)
+            throw new IllegalArgumentException("The argument is null!");
+        return node.getKey().equals(key) && node.getParents() == parents;
+    }
+
+    public String toString() {
+        return String.valueOf(key);
+    }
+
+    public T getKey() {
+        return key;
+    }
+
+    public ArrayList<DirectedGraphNode<T>> getParents() {
+        return parents;
+    }
+
+    public void addParent(DirectedGraphNode<T> parent) {
+        if(parent==null)
+            throw new IllegalArgumentException("The argument is null!");
+        this.parents.add(parent);
+    }
+
```

Comment:
In addParent() method, shouldn't we need to modify parent's child array, too?

Reply:
I was assuming that this method would always be in a "proper way", i.e. always accompanied by a call of addChild.. But you're right, as it is a public method, we can't guarantee that the user systematically does it.. So we could either just make it private (and in any case there is the public method addChild that does it well), or we could add a method addEdge(child, parent) (and respectively removeEdge(child, parent)) that would do it properly? Not sure what's best?

In any case we can't just add a child in addParent because the method addChildren uses addParent so it would loop..